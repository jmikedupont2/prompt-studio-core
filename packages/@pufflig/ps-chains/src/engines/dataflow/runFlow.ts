import { ParamValue } from "@pufflig/ps-types";
import pino from "pino";
import { Flow, RunOptions } from "../../types";
import { executionPrefix, getDefaultTargets, identity } from "./constants";
import { updateNodeInput } from "./updateNodeInput";
import { getReachableNodes } from "./utils/getReachableNodes";
import { resolveVariables } from "./utils/resolveVariables";
import { getEdgeMap, mapOutputToInput } from "./utils/utils";

const logger = pino();

/**
 * This function updates the input state of a node and executes the node with the new input.
 * Read all target nodes and update their input state. Repeat until there are no more data target nodes.
 * Nodes that are run on execution are run separately.
 *
 * @param flow
 * @param nodeId
 * @param input
 * @param runOptions
 * @returns
 */
export async function runFlow(flow: Flow, nodeId: string, input: Record<string, ParamValue>, runOptions?: RunOptions) {
  logger.level = runOptions?.logLevel || "error";

  const newState = { ...flow.state };

  // track nodes that are reachable from the current node
  const reachableNodes = getReachableNodes(nodeId, flow.definition);

  // avoid hanging on loops by keeping track of the edges that have been visited
  const visitedEdges: string[] = [];
  // track the number of times a node has been run
  const runs: Record<string, number> = {};

  const run_flow_recursive = async (nodeId: string, input: Record<string, ParamValue>) => {
    const previousState = { ...newState[nodeId] };
    const nodeConfig = flow.definition.nodes[nodeId];
    const node = flow.nodeTypes[nodeConfig?.type];

    // update node state with the input values
    const updatedFlow = { ...flow, state: newState };
    const newNodeState = await updateNodeInput(updatedFlow, nodeId, input, runOptions);
    newState[nodeId] = { ...newNodeState };
    logger.debug({ nodeId, input, newState: newState[nodeId] }, "Updated node input");

    const targets = Object.values(flow.definition.edges).filter((edge) => edge.source === nodeId);

    // guard against infinite loops
    const isAlreadyRun = (runs[nodeId] || 0) > 0;

    if (isAlreadyRun) {
      logger.debug({ nodeId }, "Node already run");
      return;
    }

    // break execution if there are no targets
    if (targets.length === 0) {
      logger.debug("No targets found");
      return;
    }

    // if the node is executable, only run it if the parent is complete
    const parentNodeId = Object.values(flow.definition.edges).find(
      (e) => e.target === nodeId && e.targetHandle.startsWith(executionPrefix)
    )?.source;
    const isParentComplete = parentNodeId ? runs[parentNodeId] > 0 : true;
    const isExecutableNode = !!node.execution?.inputs.length;

    if (!isParentComplete && isExecutableNode) {
      logger.debug("Parent node is not complete, skip execution");
      return;
    }

    // resolve variables to use during the run
    const newInput = newState[nodeId]?.input;
    let resolvedInput: Record<string, ParamValue> = {};
    try {
      logger.debug("Resolving node variables");
      resolvedInput = await resolveVariables(newInput, runOptions?.resolveReferences || (async (i) => i));
    } catch (error) {
      logger.error({ error, nodeId }, "Error resolving variables");
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    const execute = node.execute || identity;
    const getTargets = node.getTargets || getDefaultTargets(node);

    // run the node
    let result: Record<string, ParamValue> | null = null;
    try {
      // resolve references in the input
      const nodeOptions = { prevInput: previousState.input, globals: runOptions?.globals || {} };
      result = await execute(resolvedInput, nodeOptions);
      runOptions?.onNodeRunComplete?.(nodeId, result);
      logger.debug({ nodeId, result }, "Node run complete");
      // track the run
      runs[nodeId] = runs[nodeId] ? runs[nodeId] + 1 : 1;
      // break the chain if the node returns null
      if (result === null) return;
    } catch (error) {
      logger.error({ error, nodeId }, "Error running node");
      // if running the node fails, stop the chain
      runOptions?.onNodeRunError?.(nodeId, error as Error);
      return;
    }

    // ignore target nodes that are not connected to the values that have been edited
    const editedKeys = Object.keys(result);
    const dataTargetIds = targets.filter((e) => editedKeys.includes(e.sourceHandle)).map((e) => e.target);
    logger.debug({ nodeId, dataTargetIds }, "Data targets");

    // update the input of all target nodes
    for (const targetNodeId of dataTargetIds) {
      // data edges between this node and the target node
      const dataEdges = Object.values(flow.definition.edges).filter(
        (e) => e.source === nodeId && e.target === targetNodeId && !e.sourceHandle.startsWith(executionPrefix)
      );
      const edgeMap = getEdgeMap(dataEdges);
      const mappedInput = mapOutputToInput(result, edgeMap);

      // append to visited edges
      visitedEdges.push(...dataEdges.map((e) => e.id));

      // only run the node if all incoming edges have been resolved and ignore parents that are not reachable
      const incomingEdges = Object.values(flow.definition.edges).filter(
        (e) => e.target === targetNodeId && reachableNodes.has(e.source)
      );

      const areIncomingEdgesVisited = incomingEdges.every((e) => visitedEdges.includes(e.id));
      const targetConfig = flow.definition.nodes[targetNodeId];
      const targetDefinition = flow.nodeTypes[targetConfig?.type];
      const targetIsExecutable = !!targetDefinition?.execution?.inputs?.length;
      const shouldRunTarget = areIncomingEdgesVisited && !targetIsExecutable;

      if (shouldRunTarget) {
        logger.debug({ nodeId, targetNodeId, mappedInput }, "Running target node");
        await run_flow_recursive(targetNodeId, mappedInput);
      } else {
        logger.debug({ nodeId, targetNodeId, mappedInput }, "Updating target node inputs");
        const newNodeState = await updateNodeInput(updatedFlow, targetNodeId, mappedInput, runOptions);
        newState[targetNodeId] = { ...newNodeState };
      }
    }

    // run the execution target
    if (runOptions?.mode === "dataflow") return;

    // define the execution order
    const executionOrder = await getTargets(resolvedInput, result, { prevInput: previousState.input });
    const executionTargets = targets.filter(
      (edge) => edge.sourceHandle.startsWith(executionPrefix) && edge.source === nodeId
    );

    logger.debug({ executions: executionOrder }, "Execution Order");

    for (const execution of executionOrder) {
      visitedEdges.push(...executionTargets.map((e) => e.id));
      const targetId = executionTargets.find((e) => e.sourceHandle === execution.execSource)?.target;

      if (targetId) {
        // override the infinite loop guard
        runs[targetId] = runs[targetId] ? runs[targetId] - 1 : 0;

        // data edges between this node and the target node
        const dataEdges = Object.values(flow.definition.edges).filter(
          (e) => e.source === nodeId && e.target === targetId && !e.sourceHandle.startsWith(executionPrefix)
        );

        const edgeMap = getEdgeMap(dataEdges);
        const mappedInput = edgeMap ? mapOutputToInput(execution.inputs, edgeMap) : {};

        logger.debug({ nodeId, targetId, mappedInput }, "Running executable node");
        await run_flow_recursive(targetId, mappedInput);
      } else {
        logger.debug({ executionTargets, executionSource: execution.execSource }, "No target found for source");
      }
    }
  };

  await run_flow_recursive(nodeId, input);
  return newState;
}
