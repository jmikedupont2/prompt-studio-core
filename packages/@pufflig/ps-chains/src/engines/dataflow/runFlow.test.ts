import {
  configOnlyFlow,
  execWithLoop,
  loopWithJoin,
  mappedExample,
  multiInput,
  multiInputWithOutput,
  multiNodes,
  multistep,
  severalConnections,
  simpleExec,
  simpleExecWithData,
  simpleExistingState,
  simpleFlow,
  simpleFlowWithExec,
  simpleFlowWithVars,
  simpleLoop,
  singleNodeFlow,
} from "../../mocks/chains";
import { runFlow } from "./runFlow";
import { produce } from "immer";

/**
 * -> error
 */
test("should throw error if the node is undefined", async () => {
  try {
    await runFlow(singleNodeFlow, "not_a_node", { data: "Hello World" });
  } catch (err) {
    expect(err).toMatchSnapshot();
  }
});

test("should run even when lifecycle methods are missing", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    configOnlyFlow,
    "n1",
    { data: "Hello World" },
    {
      onNodeInputUpdate,
      onNodeRunError: () => {},
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1)
 */
test("run a flow with a single node", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    singleNodeFlow,
    "n1",
    { data: "Hello World" },
    {
      onNodeInputUpdate,
      onNodeRunError: () => {},
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1) -> (2)
 */
test("running a node updates the input of its children", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleFlow,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1x)   (x2x)
 *        ->  (2) -> (X)
 */
test("nodes with an execution input are not run without an execution connection", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleFlowWithExec,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeInputUpdate,
      onNodeRunComplete,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1) -> (1)
 */
test("avoid hanging on loops", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleLoop,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeInputUpdate,
      onNodeRunComplete,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1)
 */
test("merge existing state on input", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleExistingState,
    "n1",
    { data1: "IS_OVERWRITTEN" },
    {
      onNodeInputUpdate,
      onNodeRunComplete,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1a) -> (b2c) -> (a3)
 */
test("correctly map output from one node onto another", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    mappedExample,
    "n1",
    { data: "test data" },
    {
      onNodeInputUpdate,
      onNodeRunComplete,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(2);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1a) -----------> (a4)
 *       \--> (a3b) --> (b4)
 */
test("can handle input from multiple origins", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    multiInput,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(4);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(2);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1a) -----------> (a4c) --> (a5)
 *       \--> (a3b) --> (b4)
 */
test("only run a node once all inputs are resolved", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    multiInputWithOutput,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(5);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(3);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1a) ---------> (a4c) --> (a5)
 *            (X) --> (b4)
 */
test("ignore inputs from unreachable nodes", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();

  const chain = produce(multiInputWithOutput, (draftState) => {
    delete draftState.definition.edges.e1;
  });

  const res = await runFlow(
    chain,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(2);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1a) -----------------> (a4c) --> (a5)
 *            (X) --> (X) --> (b4)
 */
test("ignore inputs from nested unreachable nodes", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();

  const chain = produce(multistep, (draftState) => {
    delete draftState.definition.edges.e1;
  });

  const res = await runFlow(
    chain,
    "n1",
    { data: "Hello {{World}}" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(2);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1) -> (2)
 *         ^
 */
test("update array variables", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleFlow,
    "n2",
    {
      variables: [
        { id: "world", name: "world", type: "text", description: "", defaultValue: "test1" },
        { id: "hello", name: "hello", type: "text", description: "", defaultValue: "test2" },
      ],
    },
    {
      onNodeInputUpdate,
      onNodeRunComplete,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1*ref) -> (2)
 */
test("can resolve variables correctly", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleFlowWithVars,
    "n1",
    {},
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
      resolveReferences: async (variable) => {
        if (variable === "file:MY_FILE") {
          return "INSERTED_FILE";
        }
        return "${{ps:ref:" + variable + "}}";
      },
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * -> (1) => (2) => (3)
 */
test("run nodes connected through execution nodes", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleExec,
    "n1",
    {},
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(2);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1) => (2) => (3)
 *         ^
 */
test("do not run an executable node unless the parent was run already", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleExec,
    "n2",
    {},
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(1);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(0);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1) => (2) => (X)
 *
 */
test("when running a flow in dataflow mode, do no run child executable nodes", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    simpleExecWithData,
    "n1",
    {},
    {
      mode: "dataflow",
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1) => (🔄2) =2> (3) =2> (4)
 *
 */
test("a node can run its children multiple times", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    execWithLoop,
    "n1",
    {},
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(6);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(4);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (0) => (🔄1) =2> (2) =2> (3)
 */
test("a node can run its children multiple times 2", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    loopWithJoin,
    "n0",
    {},
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(8);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(4);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1.1) -> (2.1)
 *       -> (2.2)
 */
test("running a node with several connections to a child should update the child correctly", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    severalConnections,
    "n1",
    { data: "TEST DATA" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(3);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});

/**
 * (1.1)a -> (2.1)a
 * (1.2)x -> (2.2)b
 */
test("running a node with a missing value does not override the default value in the target node", async () => {
  const onNodeInputUpdate = jest.fn();
  const onNodeRunComplete = jest.fn();
  const onNodeRunError = jest.fn();
  const res = await runFlow(
    multiNodes,
    "n1",
    { data1: "TEST DATA" },
    {
      onNodeRunComplete,
      onNodeInputUpdate,
      onNodeRunError,
    }
  );
  expect(onNodeInputUpdate).toHaveBeenCalledTimes(2);
  expect(onNodeRunComplete).toHaveBeenCalledTimes(1);
  expect(onNodeRunError).toHaveBeenCalledTimes(0);
  expect(res).toMatchSnapshot();
});
