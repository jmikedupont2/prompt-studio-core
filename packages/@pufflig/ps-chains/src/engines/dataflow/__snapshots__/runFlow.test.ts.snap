// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`a node can run its children multiple times 1`] = `
{
  "n1": {
    "input": {
      "data": "",
    },
  },
  "n2": {
    "input": {
      "list": [
        "a",
        "b",
      ],
    },
    "status": "idle",
  },
  "n3": {
    "input": {
      "data": "b",
    },
  },
  "n4": {
    "input": {
      "data": "b",
    },
  },
}
`;

exports[`a node can run its children multiple times 2 1`] = `
{
  "n0": {
    "input": {
      "data": "",
    },
  },
  "n1": {
    "input": {
      "list": [
        "a",
        "b",
      ],
    },
    "status": "idle",
  },
  "n2": {
    "input": {
      "data": "",
      "list": [
        "a",
        "b",
      ],
    },
  },
  "n3": {
    "input": {
      "data": [
        "a",
        "b",
      ],
    },
  },
}
`;

exports[`avoid hanging on loops 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`can handle input from multiple origins 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "Hello {{World}}",
    },
  },
}
`;

exports[`can resolve variables correctly 1`] = `
{
  "n1": {
    "input": {
      "data": "\${{ps:ref:file:MY_FILE}} {{keep}}",
    },
    "status": "idle",
  },
  "n2": {
    "input": {
      "data": "INSERTED_FILE {{keep}}",
    },
  },
}
`;

exports[`correctly map output from one node onto another 1`] = `
{
  "n1": {
    "input": {
      "data": "test data",
    },
  },
  "n2": {
    "input": {
      "data1": "test data",
      "data2": "",
    },
  },
  "n3": {
    "input": {
      "data": "test data",
    },
  },
}
`;

exports[`do not run an executable node unless the parent was run already 1`] = `
{
  "n2": {
    "input": {
      "data": "",
    },
  },
}
`;

exports[`ignore inputs from nested unreachable nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n4": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "",
    },
  },
  "n5": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`ignore inputs from unreachable nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "",
    },
  },
  "n4": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`merge existing state on input 1`] = `
{
  "n1": {
    "input": {
      "data1": "IS_OVERWRITTEN",
      "data2": "KEEP",
    },
    "status": "idle",
  },
}
`;

exports[`nodes with an execution input are not run without an execution connection 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`only run a node once all inputs are resolved 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n3": {
    "input": {
      "data1": "Hello {{World}}",
      "data2": "Hello {{World}}",
    },
  },
  "n4": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`run a flow with a single node 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello World",
    },
    "status": "idle",
  },
}
`;

exports[`run nodes connected through execution nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "",
    },
  },
  "n2": {
    "input": {
      "data": "",
    },
  },
  "n3": {
    "input": {
      "data": "",
    },
  },
}
`;

exports[`running a node updates the input of its children 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
  "n2": {
    "input": {
      "data": "Hello {{World}}",
    },
  },
}
`;

exports[`running a node with a missing value does not override the default value in the target node 1`] = `
{
  "n1": {
    "input": {
      "data1": "TEST DATA",
    },
  },
  "n2": {
    "input": {
      "data1": "TEST DATA",
      "data2": "DEFAULT_DATA_2",
    },
  },
}
`;

exports[`running a node with several connections to a child should update the child correctly 1`] = `
{
  "n1": {
    "input": {
      "data": "TEST DATA",
    },
  },
  "n2": {
    "input": {
      "data1": "TEST DATA",
      "data2": "TEST DATA",
    },
  },
}
`;

exports[`should run even when lifecycle methods are missing 1`] = `
{
  "n1": {
    "input": {
      "data": "Hello World",
    },
    "status": "idle",
  },
}
`;

exports[`should throw error if the node is undefined 1`] = `[Error: Definition for node not_a_node not found]`;

exports[`update array variables 1`] = `
{
  "n2": {
    "input": {
      "data": "",
      "variables": [
        {
          "defaultValue": "test1",
          "description": "",
          "id": "world",
          "name": "world",
          "type": "text",
        },
        {
          "defaultValue": "test2",
          "description": "",
          "id": "hello",
          "name": "hello",
          "type": "text",
        },
      ],
    },
  },
}
`;

exports[`when running a flow in dataflow mode, do no run child executable nodes 1`] = `
{
  "n1": {
    "input": {
      "data": "",
    },
  },
  "n2": {
    "input": {
      "data": "",
    },
  },
}
`;
